File: cs/demo/IceStorm/clock/Publisher.cs
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

using Demo;
using System;
using System.Diagnostics;
using System.Globalization;
using System.Reflection;

[assembly: CLSCompliant(true)]

[assembly: AssemblyTitle("IceStormClockPublisher")]
[assembly: AssemblyDescription("IceStorm clock demo publisher")]
[assembly: AssemblyCompany("ZeroC, Inc.")]

public class Publisher
{
    public class App : Ice.Application
    {
        public override int run(string[] args)
        {
            string option = "None";
            string topicName = "time";
            int i;

            for(i = 0; i < args.Length; ++i)
            {
                String oldoption = option;
                if(args[i].Equals("--datagram"))
                {
                    option = "Datagram";
                }
                else if(args[i].Equals("--twoway"))
                {
                    option = "Twoway";
                }
                else if(args[i].Equals("--oneway"))
                {
                    option = "Oneway";
                }
                else if(args[i].StartsWith("--"))
                {
                    usage();
                    return 1;
                }
                else
                {
                    topicName = args[i++];
                    break;
                }

                if(!oldoption.Equals(option) && !oldoption.Equals("None"))
                {
                    usage();
                    return 1;
                }
            }

            if(i != args.Length)
            {
                usage();
                return 1;
            }

            IceStorm.TopicManagerPrx manager = IceStorm.TopicManagerPrxHelper.checkedCast(
                communicator().propertyToProxy("TopicManager.Proxy"));
            if(manager == null)
            {
                Console.WriteLine("invalid proxy");
                return 1;
            }

            //
            // Retrieve the topic.
            //
            IceStorm.TopicPrx topic;
            try
            {
                topic = manager.retrieve(topicName);
            }
            catch(IceStorm.NoSuchTopic)
            {
                try
                {
                    topic = manager.create(topicName);
                }
                catch(IceStorm.TopicExists)
                {
                    Console.WriteLine("temporary error. try again.");
                    return 1;
                }
            }

            //
            // Get the topic's publisher object, and create a Clock proxy with
            // the mode specified as an argument of this application.
            //
            Ice.ObjectPrx publisher = topic.getPublisher();
            if(option.Equals("Datagram"))
            {
                publisher = publisher.ice_datagram();
            }
            else if(option.Equals("Twoway"))
            {
                // Do nothing.
            }
            else // if(oneway)
            {
                publisher = publisher.ice_oneway();
            }
            ClockPrx clock = ClockPrxHelper.uncheckedCast(publisher);

            Console.WriteLine("publishing tick events. Press ^C to terminate the application.");
            try
            {
                while(true)
                {
                    clock.tick(System.DateTime.Now.ToString("G", DateTimeFormatInfo.InvariantInfo));
                    System.Threading.Thread.Sleep(1000);
                }
            }
            catch(Ice.CommunicatorDestroyedException)
            {
                // Ignore
            }

            return 0;
        }

        public void usage()
        {
            Console.WriteLine("Usage: " + appName() + " [--datagram|--twoway|--oneway] [topic]");
        }
    }

    public static int Main(string[] args)
    {
        App app = new App();
        return app.main(args, "config.pub");
    }
}

File: vb/demo/IceStorm/clock/Publisher.vb
' **********************************************************************
'
' Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
'
' This copy of Ice is licensed to you under the terms described in the
' ICE_LICENSE file included in this distribution.
'
' **********************************************************************

Imports System
Imports Demo

Module ClockC

    Class Publisher
        Inherits Ice.Application

        Public Overloads Overrides Function run(ByVal args() As String) As Integer
            Dim properties As Ice.Properties = communicator().getProperties()

            Dim proxyProperty As String = "IceStorm.TopicManager.Proxy"
            Dim proxy As String = properties.getProperty(proxyProperty)
            If proxy.Length = 0 Then
                Console.Error.WriteLine("property `" & proxyProperty & "' not set")
                Return 1
            End If

            Dim basePrx As Ice.ObjectPrx = communicator().propertyToProxy("IceStorm.TopicManager.Proxy")
            Dim manager As IceStorm.TopicManagerPrx = IceStorm.TopicManagerPrxHelper.checkedCast(basePrx)
            If manager Is Nothing Then
                Console.Error.WriteLine("invalid proxy")
                Return 1
            End If

	    Dim topicName as String = "time"
	    Dim datagram as Boolean = false
	    Dim twoway as Boolean = false
	    Dim optsSet as Integer = 0
	    For i As Integer = 0 To args.Length -1
		If args(i).Equals("--datagram") Then
		    datagram = true
		    optsSet = optsSet + 1
		Elseif args(i).Equals("--twoway") Then
		    twoway = true
		    optsSet = optsSet + 1
		Elseif args(i).Equals("--oneway") Then
		    optsSet = optsSet + 1
		Elseif args(i).StartsWith("--") Then
		    usage()
		    Return 1
		Else
		    topicName = args(i)
		    Exit For
		End if
    	    Next

	    If optsSet > 1 Then
		usage()
		Return 1
	    End If

	    '
	    ' Retrieve the topic.
	    '
            Dim topic As IceStorm.TopicPrx
            Try
                topic = manager.retrieve(topicName)
            Catch ex As IceStorm.NoSuchTopic
                Try
                    topic = manager.create(topicName)
                Catch e As IceStorm.TopicExists
                    Console.Error.WriteLine("temporary error. try again.")
                    Return 1
                End Try
            End Try

	    '
	    ' Get the topic's publisher object, and create a Clock proxy with
	    ' the mode specified as an argument of this application.
	    '
	    Dim publisher as Ice.ObjectPrx = topic.getPublisher()
	    If datagram Then
		publisher = publisher.ice_datagram()
	    Elseif twoway Then
		' Do nothing.
	    Else  ' if oneway
		publisher = publisher.ice_oneway()
	    End If
            Dim clock As ClockPrx = ClockPrxHelper.uncheckedCast(publisher)

            Console.Out.WriteLine("publishing tick events. Press ^C to terminate the application.")
            Try
                While 1
                    clock.tick(DateTime.Now.ToString("MM/dd/yyyy HH:mm:ss"))

                    System.Threading.Thread.Sleep(1000)
                End While
            Catch ex As Ice.CommunicatorDestroyedException
            End Try

            Return 0
        End Function

	Public Sub usage
	    Console.WriteLine("Usage: " + appName() + " [--datagram|--twoway|--oneway] [topic]")
	End Sub
    End Class

    Public Sub Main(ByVal args() As String)
        Dim app As Publisher = New Publisher
        Dim status As Integer = app.Main(args, "config.pub")
        System.Environment.Exit(status)
    End Sub
End Module
